src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -44,9 +47,173 @@ void uct_sockcm_ep_put_sock_id(uct_sockcm_ctx_t *sock_id_ctx)
+                if ((-1 == getsockopt(fd, SOL_SOCKET, SO_ERROR, 
+                                     (void *) &optval, (socklen_t *) &optlen))
+                    || (optval != 0)) {
+                    ucs_error("socket connection error");
+                    status = UCS_ERR_UNREACHABLE;
+                    ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+                    goto out;
+                }
+                ucs_debug("connecion successful\n");
+            }
+
+            status = ucs_async_modify_handler(fd, 0);
+            if (status != UCS_OK) {
+                goto out;
+            }
+
+            status = uct_sockcm_ep_send_client_info(iface, ep);
+            if (status != UCS_OK) {
+                goto out;
+            }
+
+            status = ucs_async_modify_handler(fd, UCS_EVENT_SET_EVWRITE);
+            if (status != UCS_OK) {
+                goto out;
+            }
+            return;
+        case UCT_SOCKCM_EP_CONN_STATE_INFO_SENT:
+            ucs_debug("handling send state\n");
+            status = ucs_async_modify_handler(fd, UCS_EVENT_SET_EVREAD);
+            if (status != UCS_OK) {
+                goto out;
+            }
+            /* recv to see if listener accepted or not */
+            recv_len = recv(ep->sock_id_ctx->sock_id, (char *) &conn_status,
+                            sizeof(conn_status), 0);
+            ucs_debug("sockcm_listener: recv len = %d\n", (int) recv_len);
+            if (recv_len == -1) {
+                if (errno == EAGAIN || errno == EWOULDBLOCK) {
+                    return;
+                } else {
+                   status = UCS_ERR_IO_ERROR;
+                   ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+                   goto out;
+                }
+            } else if (recv_len == 0) {
+                ucs_debug("server closed; client also closing\n");
+                status = UCS_ERR_IO_ERROR;
+                ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+                goto out;
+            }
+
+            status = conn_status ? UCS_ERR_REJECTED : UCS_OK;
+            if (UCS_OK == status) {
+                ucs_debug("event_handler OK after accept\n");
+                ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CONNECTED;
+
+                status = ucs_async_modify_handler(fd, 0);
+                goto out;
+            } else {
+                ucs_debug("event_handler REJECTED after reject\n");
+                ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+                goto out;
+            }
+            return;
+        case UCT_SOCKCM_EP_CONN_STATE_CONNECTED:
+            status = UCS_OK;
+            ucs_debug("handling connected state\n");
+            break;
+        case UCT_SOCKCM_EP_CONN_STATE_CLOSED:
+        default:
+            ucs_debug("handling closed/default state\n");
+            status = UCS_ERR_IO_ERROR;
+            break;
+    }
+
+out:
+    pthread_mutex_lock(&ep->ops_mutex);
```

Comment:
i don't see ep->ops_mutex is recursive, why are you taking it in nested fashion?

Reply:
uct_sockcm_ep_invoke_completions also takes the lock
i think this lock is not needed at all for sockcm

Reply:
@yosefe There are two main reasons why I'm using lock here.

* `uct_sockcm_ep_invoke_completions` (which is identical to `uct_rdmacm_ep_invoke_completions`) expects the caller to have grabbed the mutex lock before being called - [rdmacm](https://github.com/openucx/ucx/blob/master/src/uct/ib/rdmacm/rdmacm_ep.c#L255), [sockcm](https://github.com/openucx/ucx/blob/master/src/uct/tcp/sockcm/sockcm_ep.c#L209)
* This piece of code was pretty much derived from how `rdmacm` handles things inside the asynchronous handler. To that end, `rdmacm` uses mutex lock in the 3 locations:  
  - `ep_flush` to update `ep->status` in a thread-safe way (`ep->status` is initially `UCS_INPROGRESS` but this can change to `UCS_UNREACHABLE` or `UCS_OK` depending on failure/success and can be set by asynchronous handler) 
  - To set ep_failed and invoke completions on the endpoint in failure case 
  - To invoke completions in successful ep creation case. 

IIUC, the main thread is likely to read `ep->status` during progress but as this can be written by asynchronous thread, the mutex prevents simultaneous read and write by two separate threads. Since `sockcm` implementation of setting `ep->status`, `ep_invoke_completions`, and `ep_set_failed`  are identical to `rdmacm`, it uses its own mutex lock to control access to `ep->status` and to ensure that `ep_flush` operations are executed in a thread-safe way.

Please lmk if my understanding is wrong and that mutex usage in `sockcm` asynchronous handler is still not required.